src/t8_forest/t8_forest_cxx.cxx

```diff
@@ -2022,6 +2019,104 @@ t8_forest_print_all_leaf_neighbors (t8_forest_t forest)
+  const int element_level = scheme->t8_element_level (element);
+  /* Compute the linear id. */
+  const t8_linearidx_t element_id = scheme->t8_element_get_linear_id (element, element_level);
+  /* Search for the element.
+   * The search returns the largest index i,
+   * such that the element at position i has a smaller id than the given one.
+   * If no such i exists, it returns -1. */
+  const t8_locidx_t search_result
+    = t8_forest_bin_search_lower (elements, element_id, element_level);
+  if (search_result < 0) {
+    /* The element was not found. */
+    return 0;
+  }
+  /* An element was found but it may not be the candidate element. 
+   * To identify whether the element was found, we compare these two. */
+  const t8_element_t *check_element = t8_element_array_index_locidx (elements, search_result);
+  T8_ASSERT (check_element != NULL);
+  /* If the compare function returns 0, the elements are equal and we return true. */
+  return (scheme->t8_element_compare (element, check_element) == 0);
+}
+
+int
+t8_forest_leaf_is_boundary (const t8_forest_t forest, t8_locidx_t local_tree, const t8_element_t *leaf, int face)
+{
+  T8_ASSERT (t8_forest_is_committed (forest));
+  T8_ASSERT (t8_forest_element_is_leaf (forest, leaf, local_tree));
+  T8_ASSERT (leaf != NULL);
+
+  const t8_eclass_t tree_class = t8_forest_get_tree_class (forest, local_tree);
+  const t8_eclass_scheme_c *scheme = t8_forest_get_eclass_scheme (forest, tree_class);
+
+  /* Check whether this leaf is at the boundary of its tree. */
+  const int is_root_boundary = scheme->t8_element_is_root_boundary (leaf, face);
+
+  if (is_root_boundary) {
+    /* This leaf is at a tree's boundary. It is*/
+    const int cmesh_face = scheme->t8_element_tree_face (leaf, face);
+    const t8_cmesh_t cmesh = t8_forest_get_cmesh (forest);
+    const t8_locidx_t cmesh_local_tree = t8_forest_ltreeid_to_cmesh_ltreeid (forest, local_tree);
+    int tree_boundary = t8_cmesh_tree_face_is_boundary (cmesh, cmesh_local_tree, cmesh_face);
+    if (tree_boundary) {
+      return 1;
+    }
+  }
+
+  /* This leaf is not at the tree's boundary.
+    * If the forest has holes, we need to check whether this leaf is at an internal boundary.*/
+
+  if (!forest->incomplete_trees) {
+    /* The forest has no holes, thus the leaf cannot be a boundary leaf. */
+    return 0;
+  }
+  /* we need to compute the face neihgbors to know whether the element is a boundary element. */
+  const int is_balanced = t8_forest_is_balanced (forest);
+  int num_neighbors;
+  t8_element_t **neighbor_leaves;
+  t8_locidx_t *pelement_indices;
+  t8_eclass_scheme_c *pneigh_scheme;
+  /* The forest has holes, the leaf could lie inside a tree but its neighbor was deleted. */
+  t8_forest_leaf_face_neighbors (forest, local_tree, leaf, &neighbor_leaves, face, NULL, &num_neighbors,
+                                 &pelement_indices, &pneigh_scheme, is_balanced);
```

Comment:
For partitioned cmeshes, this function needs the ghost structure, which currently might crash: https://github.com/DLR-AMR/t8code/issues/825

Reply:
Note the ghost layer will not crash but instead is not allowed to be created in these cases (i.e. mesh with holes): #827

Reply:
I adressed this now by throwing an error and abort when the is_boundary function is used with a forest with holes.

An alternative solution would be to only print an error message and not abort - telling the user that inner boundaries cannot be identified.